src/deviceinput.cc

```diff
@@ -230,52 +202,47 @@ DeviceKeyMap::Result DeviceKeyMap::feed(const struct input_event input_events[],
 // -------------------------------------------------------------------------------------------------
 void DeviceKeyMap::resetState()
 {
-  m_pos = nullptr;
+  m_pos = &m_rootItem;
 }
 
 // -------------------------------------------------------------------------------------------------
 void DeviceKeyMap::reconfigure(const InputMapConfig& config)
 {
-  m_keymaps.resize(maxSequenceLength(config));
-
   // -- clear maps + state
   resetState();
-  for (auto& synKeyEventMap : m_keymaps) { synKeyEventMap.clear(); }
+  m_rootItem.nextMap.clear();
+  m_items.clear();
 
-  // -- fill maps
-  for (const auto& item: config)
+  // -- fill keymaps
+  for (const auto& configItem : config)
   {
-    if (!item.second.action || item.second.action->empty()) continue;
+    KeyEventItem* previous = nullptr;
+    KeyEventItem* current = &m_rootItem;
+    const auto& kes = configItem.first;
 
-    const auto& kes = item.first;
     for (size_t i = 0; i < kes.size(); ++i) {
-      m_keymaps[i].emplace(kes[i], nullptr);
-    }
-  }
-
-  // -- fill references
-  for (const auto& item: config)
-  {
-    if (!item.second.action || item.second.action->empty()) continue;
-
-    const auto& kes = item.first;
-    for (size_t i = 0; i < kes.size(); ++i)
-    {
-      const auto r = m_keymaps[i].equal_range(kes[i]);
-      if (r.first == r.second) continue;
-      auto& refobj = r.first->second;
-      if (!refobj) {
-        refobj = std::make_unique<Next>();
+      const auto& keyEvent = kes[i];
+      const auto it = std::find_if(current->nextMap.cbegin(), current->nextMap.cend(),
+      [&keyEvent](const KeyEventItem* item) {
+        return (item && item->keyEvent == keyEvent);
+      });
+
+      if (it != current->nextMap.cend()) {
+        previous = current;
+        current = *it;
       }
-
-      if (i == kes.size() - 1) { // last keyevent in seq
-        refobj->action = item.second.action;
+      else {
+        // Create new item if not found
+        m_items.emplace_back(KeyEventItem{keyEvent});
+        previous = current;
+        current = &m_items.back();
+        // link previous to current
+        previous->nextMap.push_back(current);
       }
-      else if (i+1 < m_keymaps.size()) // if not last keyevent in seq
-      {
-        const auto r = m_keymaps[i+1].equal_range(kes[i+1]);
-        if (r.first == r.second) continue;
-        refobj->next_events.emplace(&(*r.first));
+
+      // if last item in key event set
+      if (i == kes.size() - 1) {
+        current->action = configItem.second.action;
```

Comment:
Here, we are not checking If `action` is empty or not. A user can abuse it to silence some input from device by putting a 'blank' Key Sequence action (empty action) for input sequence.

I agree that this behaviour might be desirable in some cases (as we do not have `No Action` action to silence some Input Sequences).

@jahnf , do we need this behaviour?
 or 
should be check for action being non-empty before adding it to `nextMap` (and provide `No Action` action explicitly)?

Reply:
Yes you are right 👍  - depends on what the use case is .. silence some key sequences might be useful. 
But actually I had some code to check against that (or just ignore the configurations without action) but must have removed it before committing - probably was too late yesterday 😴 
I will add that simple check to ignore configuration items with empty action.

Reply:
The check against an empty shared_ptr is a must, but it is only for sanity reasons - all configs have an action. 
Additionally the last commit on the hidraw feature branch allows empty KeySequenceActions - so basically if you want to "mute" a certain key input on your device -> create an input mapping with input but an empty key sequence.

Reply:
Yeah. Hence, we do not need to implement any `No Action` action.